### ISSUE TITLE
Toggling explicit struct initialization per struct via attribute

### ISSUE THREAD

**Issue by kulkalkul** [c3lang/c3c#1605]
ZII is perfect for some tasks, but sometimes I know that given struct is always initialized. So, adding new fields to a struct definition adds possibility of breaking this guarantee as uninitialized fields can go unnoticed. It is possible to overcome this using initialization functions, but I find that it becomes necessity for such structs, makes those functions constructor-like. I also enjoy using struct initialization when possible.

As a solution, an attribute can be added for each struct initialization literal (pardon the term if wrong).

Struct definition (nothing special here):
```c
struct Foo {
  int a;
  int b;
  int c;
  int d;
}
```

Expression:
```c
@nopartial Foo {
  .a = 5,
  .b = 2,
  .d = 1,
  // compile error, as c is forgotten
}
```

Attribute at initialization location would allow same struct to be used with both ZII fields and explicit initialization, while preventing hard to follow bugs within big codebases.

**Comment by lerno** [comment]
I am not sure how you mean the initializer, that is in no way correct C3.

**Comment by kulkalkul** [comment]
```c
Foo foo = Foo { .a = 5, };
```
I'm talking about the part in right hand side

**Comment by lerno** [comment]
That is indeed the right location from a parsing point of view. I only have some doubts as to whether it should be included. C has had .foo = 5 initialization for ages, so is there some similar attribute in GCC?

**Comment by kulkalkul** [comment]
I don't know, I'm unfortunately not a C programmer 😅 It was something that annoyed me while using Odin, that's why I asked about whether something similar exists in C3. I read that C3's goal is to stay compatible with C as much as it can, that's also the theme I get from your comment above; so in that case, I agree that it'd probably feel out of place for the language

**Comment by lerno** [comment]
Adding attributes are in general fine. I just want to make sure all features are used.

**Comment by lerno** [comment]
There are two ways to ensure complete initialization:
1. Add an `init` method or `create_foo` function
2. Don't use designated initializers

It seems as if one is keen on keeping it explicit (2) would be sufficient. Why isn't it?

**Comment by kulkalkul** [comment]
I used too much Rust so I was biased towards compiler warning me. Since then I've written many lines of Odin and my programming style changed, I didn't feel the need of it in my most recent toy renderer/game, so I was mostly fine with lack of it, I think C3 is fine too. A small note, you probably know it, but Odin recently added \#all_or_none directive.
